### PR TITLE
Fix parsing of !important - take 2

### PR DESCRIPTION
**Pull request by jwilsson**
**Which issue #** if any, does this resolve?
#118 

Please check one:
- [x] New tests created for this change
- [ ] Tests updated for this change

This PR:
- [ ] Adds new API
- [ ] Extends existing API, backwards-compatible
- [ ] Introduces a breaking change
- [x] Fixes a bug

---

This should be a better approach to it than #119, the logic is mostly from core `postcss`. I'm hoping to find a way to consolidate it to one solution across both libraries.

### CODE REVIEW

File: lib/LessStringifier.js [hunk 1]
```diff
@@ -24,7 +30,7 @@ module.exports = class LessStringifier extends Stringifier {
     if (node.nodes) {
       this.block(node, name + params + important);
     } else {
-      const end = (node.raws.between || '') + important + (semicolon ? ';' : '');
+      const end = important + (semicolon ? ';' : '') + (node.raws.between || '');
```
Comment (jwilsson) [inline lib/LessStringifier.js]: I'm not 100% satisfied with this. There's one stringify test ("mixin with !important") that's failing without it so I thought I'd leave it in for the sake of discussion.

I haven't been able to find any downsides to it yet, though. But let me know what you think and we'll take it from there.

File: lib/LessStringifier.js [hunk 2]
```diff
@@ -9,7 +9,13 @@ module.exports = class LessStringifier extends Stringifier {
 
     let name = `${node.raws.identifier || '@'}${node.name}`;
     let params = node.params ? this.rawValue(node, 'params') : '';
-    const important = node.raws.important || '';
+    let important = '';
+
+    if (node.important && node.raws.important) {
```
Comment (shellscape) [inline lib/LessStringifier.js]: maybe I'm not reading the code above correctly, but in what scenario will `node.raws.important` be populated when `node.important` is falsy?

**Comment by jwilsson** [comment]
I've simplified the `!important` check but CI is failing due to something else (something in the config file I guess).

**Comment by shellscape** [comment]
So the tests are exactly what we needed. I wasn't quite sure about the direction of your solution though. I've been trying to follow the mantra that the PostCSS maintainers have shifted towards, in getting the tokens array setup before parsing/processing methods take over. It's a little different than string examination post-process. Here's my proposal to go the tokens route: https://github.com/shellscape/postcss-less/blob/d96075d0fba53b6c2c8650400d6944500c008faf/lib/LessParser.js#L116-L137

Granted it could probably use a little polish, but it passes the tests you've added on this PR, and doesn't need Stringifier modification. Thoughts?

**Comment by jwilsson** [comment]
@shellscape Cool! I'd love to learn some more about it. Have you seen any blog post, issue comment, etc. from the PostCSS maintainers? Or is the best way to learn more just reading the PostCSS code more closely?

I left one small comment on your solution, but otherwise I think it looks great! Especially since we don't  need any stringifier changes.

**Comment by shellscape** [comment]
OK, cool. I didn't want to go opening a new PR with this one pending until you had a chance to comment on the alternative direction. This stuff is no joke to work on and I'd bet you had some time into yours.

With regard to following the PostCSS methods; I gleaned that stuff from the code itself. There were some issues for the last major that went into some detail in comments about the token-first parsing deal as well. Nothing official, but just absorbed. Picked up most of that while doing the last rewrite to support v7. FWIW I'm hoping to go the same route with postcss-values-parser because it's _so much easier to maintain_ going that route. Parsers are hard!

On your suggestion, closing this one and will open a new PR on the other branch.

**Pull request closed without merging**